### PR TITLE
feat: InfluxQL planner learns how to project multiple measurements

### LINKED ISSUE #6896: InfluxQL-IOx: Multiple measurement support in `FROM` clause
Teach the InfluxQL query planner how to generate a DataFusion LogicalPlan that queries multiple measurements. 

# Prerequisites

* #6912 
  * Merging multiple measurements will require non-existent columns across measurements are correctly handled

# Background

Multiple measurements are listed in the `FROM` clause, either explicitly, by regular expression, or a combination of both:

```sql
SELECT * FROM cpu, mem, /disk/
```

Roughly, the equivalent SQL operation is to `UNION ALL` each matching table (measurement). In InfluxQL, the wildcard (`*`) will expand to all the fields and tags of the matching measurements, and produce a projection that merges the schema. Column references that don't exist in any given table are evaluated as `NULL`. The same applies when transforming column references in expressions that appear in the projection or `WHERE` clause.

The InfluxQL rewriter already performs the expansion of the wild cards in the projection list and `FROM` clause. Furthermore, the InfluxQL rewriter handles all the rules for determining the output schema, by rewriting the fields in the InfluxQL AST:

https://github.com/influxdata/influxdb_iox/blob/47b47f225be70b54adc1b846f3b280fbe533760e/influxdb_influxql_parser/src/select.rs#L35

with their explicit types (tag, integer, string, bool, etc). More specifically, the `data_type` field of the `VarRef` case will be set to the expected value, informing the planner how to process the column:

https://github.com/influxdata/influxdb_iox/blob/47b47f225be70b54adc1b846f3b280fbe533760e/influxdb_influxql_parser/src/expression/arithmetic.rs#L56

# Requirements

When InfluxQL queries over multiple measurements, the equivalent SQL would be the `UNION ALL`, selecting the data from each table. **It is not possible to combine, join or aggregate data across measurements in any way using InfluxQL, even through the use of subqueries.**

* The `WHERE`, `LIMIT`,`OFFSET` and `ORDER BY` clauses are applied to each `SELECT`
* For each `SELECT` in the `UNION ALL`, any columns in the projection that do not exist, should project a `NULL` value
* Any tags or fields in the original `WHERE` clause that don't exist for a given table would evaluate to `false`
  * **Optimisation**: these tables could be excluded from the logical plan completely
* The first column for each select should be the table name (measurement name in InfluxQL speak), to distinguish them in the output and to facilitate rendering them correctly in `influxqld`[^2]
* Any explicit measurements in the `FROM` clause that do not exist should be excluded from the logical plan

## Examples[^1]

These examples show how the output schema and how fields and tags are combined across measurements.

### Example 1

The `host` tag exists for both the `disk` and `cpu` measurements. The `usage_idle` field exists for the `cpu` measurement and the `free` exists for the `disk` measurement.

```sql
select host, cpu, mode, usage_idle, free FROM cpu, disk LIMIT 3;
```
```
name: cpu
time                 host                            cpu       mode usage_idle        free
----                 ----                            ---       ---- ----------        ----
2020-06-11T16:52:00Z Andrews-MBP.hsd1.ma.comcast.net cpu-total      97.93916821849783
2020-06-11T16:52:00Z Andrews-MBP.hsd1.ma.comcast.net cpu0           89.56262425447316
2020-06-11T16:52:00Z Andrews-MBP.hsd1.ma.comcast.net cpu1           99.8015873015873

name: disk
time                 host                            cpu mode usage_idle free
----                 ----                            --- ---- ---------- ----
2020-06-11T16:51:50Z Andrews-MBP.hsd1.ma.comcast.net     ro              882941362176
2020-06-11T16:51:50Z Andrews-MBP.hsd1.ma.comcast.net     rw              882941362176
2020-06-11T16:51:50Z Andrews-MBP.hsd1.ma.comcast.net     rw              882941362176
```

### Example 2: Conflicting column types

It is possible that columns across tables share the same name, but are different types (tag vs field) or field data types (float, integer, bool, etc). 

For example, given the following input data, noting that the column names are different types across the two measurements:

```
insert merge_00,col0=val00 col1=3.5,col2=false,col3="a string"
insert merge_01,col1=val01 col0=9.5,col3=true,col2="another string"
```

The following statement, which selects all the fields and tags:

```sql
> select * from merge_00, merge_01
```

Is rewritten as

```sql
SELECT col0::float AS col0, col0::tag AS col0_1, col1::float AS col1, col1::tag AS col1_1, col2::string AS col2, col3::string AS col3 FROM merge_00, merge_01
```

The planner can use the data types in the rewritten AST to inform how the logical plan is generated.

InfluxDB OSS 1.8, produces the following output, noting that the boolean field values are rendered as NULL as they are not coalesced to strings:

```
name: merge_00
time                        col0 col0_1 col1 col1_1 col2 col3
----                        ---- ------ ---- ------ ---- ----
2023-02-07T23:18:56.997745Z      val00  3.5              a string

name: merge_01
time                        col0 col0_1 col1 col1_1 col2           col3
----                        ---- ------ ---- ------ ----           ----
2023-02-07T23:19:31.148362Z 9.5              val01  another string
```

Equally, if the merged schema combined numeric and string fields, the string fields would be NULL.

[^1]: Examples use the [`metrics.lp`](https://github.com/influxdata/influxdb_iox/blob/main/test_fixtures/lineproto/metrics.lp) test data unless otherwise stated.
[^2]: A separate issue will be created to discuss how this should work

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #6896

Teach the InfluxQL planner how to generate a logical plan that selects from zero to many measurements.

# Background

The most significant user-facing change is the addition of the `iox::measurement` column, which is added as a projection to each source table, and specifies a literal which is the name of the table. This is required to distinguish which measurement each row is associated, and will be used by downstream clients to produce the correct output by grouping by measurement. It forms part of the _group key_. Tags specified in a `GROUP BY` clause would also be projected and used to group the InfluxQL output. For example, the following query, which selects from the two measurements, `cpu` and `disk`, and executed against InfluxDB v1.8:

```sql
SELECT usage_idle, free FROM cpu, disk WHERE cpu = 'cpu-total' OR device = 'disk1s1'
```

produces the following output, noting that it is grouped by the measurement name:

```
name: cpu
time                usage_idle        free
----                ----------        ----
1591894320000000000 97.93916821849783
1591894330000000000 97.70610663166448

name: disk
time                usage_idle free
----                ---------- ----
1591894310000000000            882941362176
1591894320000000000            882941370368
```

Tag columns specified in the `GROUP BY` clause will also need to be annotated to inform the client how to group the output.

## Identifying grouping fields

We would prefer to a be able to annotate each of the special columns to inform the InfluxQL client what columns are the group key. We were unable to persist any metadata through the DFSchema / Arrow schema after optimisation and planning had completed.

### Unable to add metadata to logical planner nodes

Note the following, where the match arm here doesn’t include the schema field from the `Projection` struct (which is the root of the problem):

https://github.com/apache/arrow-datafusion/blob/b673a6187ff94c604a33809e228993652b6635c6/datafusion/core/src/physical_plan/planner.rs#L724

The projection is constructed here:

https://github.com/apache/arrow-datafusion/blob/b673a6187ff94c604a33809e228993652b6635c6/datafusion/core/src/physical_plan/planner.rs#L771-L774

which constructs the new schema, and because it only references the `input_schema`, will always return the default value given the `iox::measurement` column does not exist in the table’s `input_schema` here:

https://github.com/apache/arrow-datafusion/blob/b673a6187ff94c604a33809e228993652b6635c6/datafusion/core/src/physical_plan/projection.rs#L82

## New logical and physical planner nodes

See proposal #7062 to discuss viability

### CODE REVIEW

File: iox_query/src/plan/influxql/planner.rs [hunk 1]
```diff
@@ -1120,43 +1198,44 @@ mod test {
         #[test]
         fn test_single_measurement() {
             assert_snapshot!(plan("SELECT f64_field FROM data"), @r###"
-            Projection: data.time, data.f64_field AS f64_field [time:Timestamp(Nanosecond, None), f64_field:Float64;N]
-              Sort: data.time ASC NULLS LAST [TIME:Boolean;N, bar:Dictionary(Int32, Utf8);N, bool_field:Boolean;N, f64_field:Float64;N, foo:Dictionary(Int32, Utf8);N, i64_field:Int64;N, mixedCase:Float64;N, str_field:Utf8;N, time:Timestamp(Nanosecond, None), with space:Float64;N]
+            Sort: data.time ASC NULLS LAST [iox::measurement:Dictionary(Int32, Utf8), time:Timestamp(Nanosecond, None), f64_field:Float64;N]
```
Comment (crepererum) [inline iox_query/src/plan/influxql/planner.rs:1201]: Just picking a random instance here:

I would like to see the phys. plan of this. I think we can get around a full sort and just `SortPreservingMergeExec` on the phys. layer (which IIRC DF should insert automatically) IF we tell DF that the `Projection`s actually create a sorting by that constant value. I think this is currently not possible, but we could introduce an `AssumeSortedBy` phys. node.

@alamb what do you think?

Comment (crepererum) [reply iox_query/src/plan/influxql/planner.rs:1201]: Thinking about this more: I think DF could even be smart enough (in some mid-term future) to infer that adding constant columns via `ProjectionExec` means the data is sorted by that particular column.

Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner.rs:1201]: @crepererum I created #7062, which includes a physical plan showing the sort.

Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner.rs:1201]: > Thinking about this more: I think DF could even be smart enough (in some mid-term future) to infer that adding constant columns via `ProjectionExec` means the data is sorted by that particular column.

That would be ideal

Comment (alamb) [reply iox_query/src/plan/influxql/planner.rs:1201]: > That would be ideal

Yes, that would be very cool


Maybe we just have to add the relevant column to  https://docs.rs/datafusion/18.0.0/datafusion/physical_plan/trait.ExecutionPlan.html#tymethod.output_ordering

Though we would have to ensure it didn't mess up other optimizations. I think there is already some additional sophistication in equivalence_properties that probably covres it


https://docs.rs/datafusion/18.0.0/datafusion/physical_plan/trait.ExecutionPlan.html#method.equivalence_properties

**Comment by alamb** [comment]
> I would like to see the phys. plan of this. I think we can get around a full sort and just SortPreservingMergeExec on the phys. layer (which IIRC DF should insert automatically) IF we tell DF that the Projections actually create a sorting by that constant value. I think this is currently not possible, but we could introduce an AssumeSortedBy phys. node.
> 
> @alamb what do you think?


I think this is an excellent idea

**Review by alamb — APPROVED** [review]
Avoiding the sort would be great, if possible, as @crepererum  says. However, I also think it would probably be fine to treat this as a optimization as the sort will be correct

File: iox_query/src/plan/influxql/planner.rs [hunk 2]
```diff
@@ -149,63 +151,101 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
...
-            )),
-        }?;
+        let Some(plan) = plans.pop_front() else { return LogicalPlanBuilder::empty(false).build(); };
+        let plan = self.project_select(plan, select)?;
 
-        let schemas = Schemas::new(plan.schema())?;
-        let tz = select.timezone.map(|tz| *tz);
+        // If there are multiple measurements, we need to sort by the measurement column
+        // NOTE: Ideally DataFusion would maintain the order of the UNION ALL, which would eliminate
+        //  the need to sort by measurement.
```
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply iox_query/src/plan/influxql/planner.rs:167]: Done

**Pull request merged**